### PR TITLE
Fix errors from issue 5970

### PR DESCRIPTION
**Pull request by thomasspriggs**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/smt2/smt2_conv.cpp [hunk 1]
```diff
@@ -741,13 +750,19 @@ literalt smt2_convt::convert(const exprt &expr)
...
   out << "; convert\n";
   out << "(define-fun ";
   convert_literal(l);
   out << " () Bool ";
-  convert_expr(prepared_expr);
+  if(can_cast_expr<forall_exprt>(prepared_expr))
+  {
+    out << "true";
+  }
+  else
```
Comment (TGWDB) [inline src/solvers/smt2/smt2_conv.cpp]: Should there also be one for `exists_exprt`?

File: src/solvers/smt2/smt2_conv.cpp [hunk 2]
```diff
@@ -741,13 +750,19 @@ literalt smt2_convt::convert(const exprt &expr)
   exprt prepared_expr = prepare_for_convert_expr(expr);
 
   literalt l(no_boolean_variables, false);
+
```
Comment (TGWDB) [inline src/solvers/smt2/smt2_conv.cpp]: Maybe don't add newlines?

**Review by TGWDB — CHANGES_REQUESTED** [review]

**Comment by thomasspriggs** [comment]
The original approach in this PR was fairly unsound. It only produced better results for one specific test case and it introduced new problems. I have now pushed a new approach based on ignoring errors from `z3` and inferring the values we couldn't get based the assertions also present in the SMT2 file. This should produce better results when solving with `z3`, where the quantifiers are inside a call to `cprover_assume`. However we do not have an equivalent fix for where the quantifiers are inside an assert.

File: src/solvers/smt2/smt2_conv.cpp [hunk 3]
```diff
@@ -745,6 +745,16 @@ literalt smt2_convt::convert(const exprt &expr)
...
+  out << "; Converting var_no " << l.var_no() << " with expr ID of "
+      << expr.id_string() << "\n";
+  // We're converting the expression, so store it in the defined_expressions
+  // store and in future we use the literal instead of the whole expression
+  // Note that here we are always converting, so we do not need to consider
+  // other literal kinds, only "|B###|"
+  std::stringstream tmpstr;
+  tmpstr << "|B" << l.var_no() << "|";
+  defined_expressions[expr] = tmpstr.str();
+  // End temporary storage
```
Comment (TGWDB) [inline src/solvers/smt2/smt2_conv.cpp]: This line of comment no longer correct.

File: src/solvers/smt2/smt2_conv.cpp [hunk 4]
```diff
@@ -4314,7 +4324,28 @@ void smt2_convt::set_to(const exprt &expr, bool value)
...
+  {
+    convert_expr(prepared_expr);
+  }
+  if(!value)
+  {
+    out << ")";
+  }
+  out << ")\n";
+  return;
+  // OLD BELOW
```
Comment (TGWDB) [inline src/solvers/smt2/smt2_conv.cpp]: If this fix stays in then we should remove the old code below.

File: src/solvers/smt2/smt2_conv.h [hunk 5]
```diff
@@ -225,6 +225,7 @@ class smt2_convt : public stack_decision_proceduret
 
   typedef std::map<exprt, irep_idt> defined_expressionst;
   defined_expressionst defined_expressions;
+  std::unordered_map<irep_idt, bool> set_values;
```
Comment (TGWDB) [inline src/solvers/smt2/smt2_conv.h:231]: Doxygen this?

**Review by TGWDB — APPROVED** [review]
Approving as all changes are only tidying up/improving commentary.

File: src/solvers/smt2/smt2_conv.cpp (hunk 4 above)
Comment (thomasspriggs) [reply src/solvers/smt2/smt2_conv.cpp]: Removed. Why didn't you just remove this to begin with? :satisfied:

Comment (thomasspriggs) [reply src/solvers/smt2/smt2_conv.cpp]: I have refactored the code and removed this comment.

File: src/solvers/smt2/smt2_conv.h (hunk 5 above)
Comment (thomasspriggs) [reply src/solvers/smt2/smt2_conv.h:231]: Done.

**Pull request merged**